src/util/parallel/data.ml

```diff
@@ -0,0 +1,242 @@
+open Batteries
+
+module type DefaultType = sig
+  type t
+  val default: unit -> t
+  val to_string: t -> string
+end
+
+module ConcurrentBucket (Key: Hashtbl.HashedType) (Val: DefaultType) = struct
+  type t = {
+    key: Key.t;
+    value: Val.t Atomic.t;
+    next: t option Atomic.t;
+  }
+
+  let create key = {
+    key = key;
+    value = Atomic.make @@ Val.default ();
+    next = Atomic.make None;
+  }
+
+  let create_with_value key value = {
+    key = key;
+    value = value;
+    next = Atomic.make None;
+  }
+
+  let create_with_value_and_next key value next = {
+    key = key;
+    value = value;
+    next = Atomic.make (Some next);
+  }
+
+  let find_option sll key =
+    let rec find sll key =
+      if Key.equal sll.key key then Some sll.value
+      else match Atomic.get sll.next with
+        | None -> None
+        | Some next -> find next key
```

Comment:
`Option.map` ?

Reply:
`Option.map` is `('a -> 'b) -> 'a option -> 'b option` , but `find` already returns a `'b option` on the then branch. Or am I missing a simple way to rewrite find's then branch?

Reply:
Nah, I didn't look close enough. `Option.bind` then.

Reply:
Yup it works, did not know it existed.